libyul/optimiser/ForLoopConditionIntoBody.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/ForLoopConditionIntoBody.h>
+#include <libyul/AsmData.h>
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+void ForLoopConditionIntoBody::operator()(Block& _block)
+{
+	for (Statement& _stmt: _block.statements)
+	{
+		if (_stmt.type() == typeid(ForLoop))
+		{
+			auto& forLoop = boost::get<ForLoop>(_stmt);
+			(*this)(forLoop.pre);
+			(*this)(forLoop.body);
+			(*this)(forLoop.post);
+			langutil::SourceLocation loc = locationOf(*forLoop.condition);
+			forLoop.body.statements.insert(
+				forLoop.body.statements.begin(),
+				If {
+					loc,
+					make_unique<Expression>(
+						FunctionalInstruction {
+							loc,
+							eth::Instruction::ISZERO,
+							make_vector<Expression>(std::move(*forLoop.condition))
+						}
+					),
+					Block {loc, make_vector<Statement>(Break{{}})}
+				}
+			);
+			forLoop.condition = make_unique<Expression>(
+				Literal {
+					loc,
+					LiteralKind::Number,
+					"1"_yulstring,
+					{}
+				}
+			);
+		}
+		else
+			visit(_stmt);
```

Comment:
Can't this be moved to the beginning of the for loop, unconditionally?

Reply:
Oh, I guess yes. And maybe `ForLoopConditionIntoBody::operator()(ForLoop& _floop)` instead of `ForLoopConditionIntoBody::operator()(Block& _block)`

Reply:
Updated